__tests__/classes/MembershipExpirationProcessor.js

```diff
@@ -0,0 +1,154 @@
+      const expirationProcessor = setup()
+
+      await generateMemberAndPayment({
+        memberOpts: {
+          id: 11,
+          membershipEndDate: addDays(fromDate, 27)
+        }
+      })
+
+      await expirationProcessor.processMemberships(fromDate, addDays(fromDate, 30))
+      const updatedMember = ah.api.models.member.findOne({ where: { id: 11 } })
+
+      expect(updatedMember.notifiedForExpiringDate).toEqual(updatedMember.membershipEndDate)
+    })
+
+    test('do not notify expiring member with email when already notified ', async () => {
+      const expirationProcessor = setup()
+      const sendMailSpy = jest.spyOn(expirationProcessor, 'enqueueSendMail')
+
+      await generateMemberAndPayment({
+        memberOpts: {
+          id: 11,
+          membershipEndDate: addDays(fromDate, 27),
+          notifiedForExpiringDate: addDays(fromDate, 27)
+        }
+      })
+
+      await expirationProcessor.processMemberships(fromDate, addDays(fromDate, 30))
+
+      expect(sendMailSpy).toHaveBeenCalledTimes(0)
+    })
+
+    test('do not notify expiring member with payment.paymentType == "group" ', async () => {
+      const expirationProcessor = setup()
+      const sendMailSpy = jest.spyOn(expirationProcessor, 'enqueueSendMail')
+
+      await generateMemberAndPayment({
+        memberOpts: {
+          id: 11,
+          membershipEndDate: addDays(fromDate, 27)
+        },
+        paymentOps: {
+          paymentType: 'group'
+        }
+      })
+
+      await expirationProcessor.processMemberships(fromDate, addDays(fromDate, 30))
+
+      expect(sendMailSpy).toHaveBeenCalledTimes(0)
+    })
+
+    test('do not notify expiring members who are not billing member', async () => {
+      const expirationProcessor = setup()
+      const sendMailSpy = jest.spyOn(expirationProcessor, 'enqueueSendMail')
+      await ah.api.models.member.create(await generateMember({ id: 111, membershipEndDate: addDays(fromDate, 27) }))
+      await generateMemberAndPayment({
+        memberOpts: {
+          id: 11,
+          membershipEndDate: addDays(fromDate, 27)
+        },
+        paymentOps: {
+          members: [11, 111],
+          billingMemberId: 111
+        }
+      })
+
+      await expirationProcessor.processMemberships(fromDate, addDays(fromDate, 30))
+
+      expect(sendMailSpy).toHaveBeenCalledTimes(1)
+      expect(sendMailSpy).toHaveBeenCalledWith(expect.objectContaining({ id: 111 }))
+    })
+
+    test('throw error when from date is not provided ', async () => {
+      const expirationProcessor = setup()
+      expect(await expirationProcessor.processMemberships(null, new Date())).toEqual(expectedError)
+    })
+
+    test('throw error when from date is not provided ', async () => {
+      const expirationProcessor = setup()
+      expect(await expirationProcessor.processMemberships(new Date(), null)).toEqual(expectedError)
+    })
+  })
+})
```

Comment:
Missing cases:

1. Member without payments -> no email
2. Member with older regular payment and newer group payment -> no email
3. Member with newer regular payment and older group payment -> send
4. Member with family payment but also is billingMember -> send
5. Member with expired membership (i.e. membershipEndDate < startDate) -> no email